lib/pbench/server/api/resources/datasets_list.py

```diff
@@ -139,21 +159,96 @@ def _get(
+                        native_key = keys.pop(0).lower()
+                    else:
+                        try:
+                            c = getattr(Dataset, second)
+                            column = (
+                                c
+                                if c.type.python_type is str
+                                else cast(getattr(Dataset, second), String)
+                            )
+                        except AttributeError as e:
+                            raise APIAbort(
+                                HTTPStatus.BAD_REQUEST, str(MetadataBadKey(k))
+                            ) from e
+                        use_dataset = True
+                        terms = [column.contains(v) if contains else column == v]
+                elif native_key == Metadata.USER:
+                    if not user_id:
+                        raise APIAbort(
+                            HTTPStatus.UNAUTHORIZED,
+                            f"Metadata key {k} cannot be used by an unauthenticated client",
+                        )
+                    user_private = [Metadata.user_id == user_id]
+
+                if not use_dataset:
+                    expression = Metadata.value[keys].as_string()
+                    terms = [
+                        Metadata.key == native_key,
+                        expression.contains(v) if contains else expression == v,
+                    ]
+                    if user_private:
+                        terms.extend(user_private)
+                filter = and_(*terms)
+
+                if combine_or:
+                    or_list.append(filter)
+                else:
+                    if or_list:
+                        and_list.append(or_(*or_list))
+                        or_list.clear()
+                    and_list.append(filter)
+
+            if or_list:
+                and_list.append(or_(*or_list))
+
+            query = query.filter(and_(*and_list))
+
         query = self._build_sql_query(json.get("owner"), json.get("access"), query)
 
-        # Useful for debugging, but verbose: this displays the fully expanded
-        # SQL `SELECT` statement.
-        if current_app.logger.isEnabledFor(logging.DEBUG):
-            current_app.logger.debug(
-                "QUERY {}",
-                query.statement.compile(compile_kwargs={"literal_binds": True}),
+        try:
+            datasets, paginated_result = self.get_paginated_obj(
+                query=query, json=json, url=request.url
             )
-
-        # Execute the filtered query, sorted by dataset name so we have a
-        # consistent and reproducible output to compare.
-        datasets, paginated_result = self.get_paginated_obj(
-            query=query, json=json, url=request.url
-        )
+        except (AttributeError, ProgrammingError, StatementError) as e:
+            raise APIInternalError(
+                f"Constructed SQL for {json} isn't executable"
+            ) from e
```

Comment:
Yeah, the problem is that we may not be able to: it might fail if the engine really can't compile the expression. While that's unlikely now that I've "got it" it'll remain a potential issue. (That is, I did originally try to dump the SQL here, and ended up preventing getting a reasonable message at all because the logger formatter blew up...)